src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,163 @@
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            newNode.next = prev.next;
+            prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue;
+        if (index == 0) {
+            removedValue = head.value;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedValue = prev.next.value;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                tail = prev;
+            }
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        if (head == null) {
+            return false;
+        }
+
+        if (object == null) {
+            if (head.value == null) {
+                remove(0);
+                return true;
+            }
+        } else if (object.equals(head.value)) {
+            remove(0);
+            return true;
+        }
+
+        Node<T> current = head;
+        while (current.next != null) {
+            if (current.next.value == null && object == null) {
+                current.next = current.next.next;
+                if (current.next == null) {
+                    tail = current;
+                }
+                size--;
+                return true;
+            } else if (current.next.value != null && current.next.value.equals(object)) {
+                current.next = current.next.next;
+                if (current.next == null) {
+                    tail = current;
+                }
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
```

Comment:
Consider using the `unlink(Node node)` method that you will create to handle the logic of removing a node. This will help to avoid code duplication and ensure that all links are properly removed when a node is deleted.